File: Android/app/src/main/java/com/talka/fancybridge/Manager/StateManager.java
package com.talka.fancybridge.Manager;

import android.app.AlertDialog;
import android.app.Service;
import android.content.Context;
import android.content.DialogInterface;
import android.provider.Settings;
import android.telephony.TelephonyManager;

import com.talka.fancybridge.Views.AlertView;
import com.talka.fancybridge.Views.Components.CardView;
import com.talka.fancybridge.Views.LoadingView;

import static java.lang.System.exit;
import static java.lang.Thread.sleep;

public class StateManager implements StreamManager.StreamManagerListener {

    static StateManager instance = null;

    public enum GameState {
        Connected,
        Wait,
        Call,
        CallChoosePartner,
        Play;

        public static GameState fromInteger(int x) {
            switch(x) {
                case -1:
                    return Connected;
                case 0:
                    return Wait;
                case 1:
                    return Call;
                case 2:
                    return CallChoosePartner;
                case 3:
                    return Play;
            }
            return null;
        }
    }

    enum ConnectState {
        S("GameState"),
        M("Message"),
        W("Wait"),
        T("ThreeMode"),
        N("Names"),
        H("Hand"),
        C("Call"),
        P("Play"),
        D("Disconnect"),
        R("Recover");

        private final String context;

        ConnectState(final String context) {
            this.context = context;
        }

        /* (non-Javadoc)
         * @see java.lang.Enum#toString()
         */
        @Override
        public String toString() {
            return context;
        }
    }

    enum RecoverState {
        cards,
        names,
        callRecord,
        playRecord,
        boutsWin,
        done;

        public static RecoverState fromInteger(int x) {
            switch(x) {
                case 1:
                    return cards;
                case 2:
                    return names;
                case 3:
                    return callRecord;
                case 4:
                    return playRecord;
                case 5:
                    return boutsWin;
                case 6:
                    return done;
            }
            return null;
        }
    }

    public enum PlayState {
        Normal,
        LastCard,
        GameOver
    }

    public class PlayerInfo {
        public String name;
        public int turnIndex;

        public PlayerInfo(String setName, int setTurnIndex) {
            name = setName;
            turnIndex = setTurnIndex;
        }
    }

    public interface StateManagerListener {
        void updateWaitingUI(int normalCount,int threeModeCount);

        void changeStateUI(GameState state);
    }

    public interface StateManagerCallListener {
        void updateCallingUI(int index);

        void showThreeModeUI(int index);
    }

    public interface StateManagerPlayListener {
        void updatePlayingUI(int poker, int type, int lastUser);
        void recoverPlayingUI();
    }

    public Context context;
    public StateManagerListener listener;
    public StateManagerCallListener callListener;
    public StateManagerPlayListener playListener;
    StreamManager streamManager;
    public PlayerInfo playInfo = new PlayerInfo("", -1);
    public String[] players;
    public String[] threeModePlayers;
    GameState gameState = GameState.Wait;
    public Boolean isGameOver = false;

    public static synchronized StateManager getInstance() {
        if (instance == null) {
            instance = new StateManager();
        }
        return instance;
    }

    private StateManager() {
        streamManager = StreamManager.getInstance();
        streamManager.listener = this;
    }

    public void setContext(Context context) {
        this.context=context;
    }

    //MARK: - Functions
    public void reset() {
        players = null;
        gameState = GameState.Wait;
        isGameOver = false;
        if (listener != null) {
            listener.changeStateUI(gameState);
        }
    }

    public void connectServer() {
        streamManager.connect();
    }

    public void joinThreeMode() {
        streamManager.sendMessage("W");
    }

    public void choosedPartner(int choosedIndex) {
        streamManager.sendMessage(String.format("T%d", choosedIndex));
    }

    public void interruptConnect() {
        try {
            streamManager.disconnect();
        } catch (Throwable throwable) {
            throwable.printStackTrace();
        }
    }

    public void test(String text) {
        streamManager.sendMessage(text);
    }

    @Override
    public void getData(String data) {
        PokerManager pokerManager = PokerManager.getInstance();
        ConnectState connectState = ConnectState.valueOf(data.substring(0, 1));

        String info = data.substring(1);

        switch (connectState) {
            case S: {
                String[] splitArray = info.split(",");
                GameState tempState = GameState.fromInteger(Integer.parseInt(splitArray[0]));
                gameState = tempState;

                switch (gameState) {
                    case Connected: {
//                        TelephonyManager tm = (TelephonyManager) context.getSystemService(Service.TELEPHONY_SERVICE);
                        String deviceId = Settings.Secure.getString(context.getContentResolver(), Settings.Secure.ANDROID_ID);
                        streamManager.sendMessage(String.format("N%s,A-%s",playInfo.name,deviceId));
                        break;
                    }
                    case Wait: {
                        playInfo.turnIndex = Integer.parseInt(splitArray[1]);
                        break;
                    }
                    case Call: {
                        int threeModeInt = Integer.parseInt(splitArray[2]);
                        pokerManager.turnIndex = Integer.parseInt(splitArray[1]);
                        pokerManager.threeMode = (threeModeInt == 1 ? true : false);
                        break;
                    }
                    case CallChoosePartner:
                        break;
                    case Play: {
                        int lastUser = Integer.parseInt(splitArray[3]);
                        pokerManager.trump = Integer.parseInt(splitArray[1]);
                        pokerManager.turnIndex = Integer.parseInt(splitArray[2]);
                        pokerManager.callsRecord.get(lastUser).add(-1);

                        if (pokerManager.threeMode) {
                            lastUser = 1;
                        }

                        if (lastUser == playInfo.turnIndex || lastUser == (playInfo.turnIndex + 2) % 4) {
                            pokerManager.winNumber = PokerManager.totalSum - (pokerManager.trump / 7 + 7);
                        } else {
                            pokerManager.winNumber = (pokerManager.trump / 7 + 7);
                        }
                        break;
                    }
                }

                if (listener != null) {
                    listener.changeStateUI(gameState);
                }

                break;
            }
            case M: {
//                if (listener != null) {
//                    listener.updateWaitingUI(info);
//                }
                break;
            }
            case W: {
                String[] splitArray = info.split(",");
                int normalCount = Integer.parseInt(splitArray[0]);
                int threeModeCount = Integer.parseInt(splitArray[1]);

                if (listener != null) {
                    listener.updateWaitingUI(normalCount, threeModeCount);
                }
                break;
            }
            case T: {
                int calledIndex = Integer.parseInt(info);
                gameState = GameState.CallChoosePartner;
                pokerManager.turnIndex = calledIndex;
                if (callListener != null) {
                    callListener.showThreeModeUI(calledIndex);
                }
                break;
            }
            case N: {
                if( !pokerManager.threeMode ) {
                    players = info.split(",");
                }
                else {
                    threeModePlayers = info.split(",");

                    for(int i=0;i<4;++i) {
                        if( threeModePlayers[i].equals(playInfo.name) ) {
                            playInfo.turnIndex = i;
                            AlertView.showViewSetText(context,String.format("You are player %d",i+1));
                        }

                        if ( threeModePlayers[i].equals("-")) {
                            pokerManager.comIndex = i;
                        }
                    }
                }
                break;
            }
            case H: {
                if( !pokerManager.threeMode ) {
                    pokerManager.setCards(info.split(","));
                }
                else {
                    pokerManager.setOtherCards(info.split(","));
                }
                break;
            }
            case C: {
                String[] splitArray = info.split(",");
                int nowTurn = Integer.parseInt(splitArray[1]);
                int trump = Integer.parseInt(splitArray[2]);
                int lastTurn = Integer.parseInt(splitArray[0]);

                pokerManager.turnIndex = nowTurn;
                if( playInfo.turnIndex == nowTurn) {
                    AlertView.showViewSetText(context,"Your Turn");
                }
                pokerManager.callsRecord.get(lastTurn).add(trump);

                if (callListener != null) {
                    callListener.updateCallingUI(trump);
                }

                break;
            }
            case P: {
                String[] splitArray = info.split(",");
                int nextTurn = Integer.parseInt(splitArray[2]);
                int lastTurn = (nextTurn + 4 - 1) % 4;
                int poker = Integer.parseInt(splitArray[0]);
                pokerManager.turnIndex = nextTurn;

                PlayState playState = PlayState.Normal;
                if (splitArray[1].equals("0")) {
                    playState = PlayState.Normal;
                } else if (splitArray[1].equals("1")) {
                    playState = PlayState.LastCard;
                } else if (splitArray[1].equals("2")) {
                    playState = PlayState.GameOver;
                }

                switch (playState) {
                    case Normal:
                        if (poker != 0) {
                            if (pokerManager.currentFlower == -1) {
                                pokerManager.currentFlower = (poker - 1) / 13;
                            }

                            int flower = (poker - 1) / 13;
                            pokerManager.playsRecord.get(lastTurn).add(poker);
                            pokerManager.flowerCountRecord[flower] += 1;
                        }
                        else {
                            pokerManager.boutsWinRecord.add(nextTurn);
                        }
                        break;
                    case LastCard:
                        int flower = (poker - 1) / 13;
                        pokerManager.playsRecord.get(lastTurn).add(poker);
                        pokerManager.flowerCountRecord[flower] += 1;

                        if (playInfo.turnIndex == 0) {
                            new Thread(new Runnable() {
                                @Override
                                public void run() {
                                    try {
                                        sleep(2000);
                                        streamManager.sendMessage("P-1,0");
                                    } catch (InterruptedException e) {
//                                        Log.d("error");
                                    }
                                }
                            }).start();
                        }

                        pokerManager.currentFlower = -1;
                        pokerManager.turnIndex = -1;

                        break;
                    case GameOver:
                        pokerManager.boutsWinRecord.add(nextTurn);
                        isGameOver = true;
                        break;
                }

                if (playListener != null) {
                    playListener.updatePlayingUI(poker, Integer.parseInt(splitArray[1]), lastTurn);
                }

                break;
            }
            case D: {
                if(StateManager.getInstance().isGameOver) {
                    return;
                }

                if(info.equals("0")) {
                    AlertDialog.Builder alert = new AlertDialog.Builder(context);
                    alert.setTitle("Server Message");
                    alert.setMessage("Someone Disconnect");
                    alert.setPositiveButton("OK", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int i) {
                        }
                    });
                    alert.show();
                    LoadingView.start(context);
                }
                else if(info.equals("1")) {
                    if((gameState == GameState.CallChoosePartner && pokerManager.turnIndex == playInfo.turnIndex)
                            || gameState != GameState.CallChoosePartner) {
                        LoadingView.stop(context);
                    }
                }
                else {
                    closeConnectAlert("close connect");
                }

                break;
            }
            case R: {
                String stateStr = info.substring(0, 1);
                RecoverState recoverState = RecoverState.fromInteger(Integer.parseInt(stateStr));
                String recoverInfo = info.substring(2);

                switch (recoverState) {
                    case cards: {
                        String[] splitArray = recoverInfo.split("\\^");
                        String cardsStr = splitArray[0];
                        pokerManager.setCards(cardsStr.split(","));
                        if(splitArray.length > 1) {
                            String otherCardsStr = splitArray[1];
                            pokerManager.setOtherCards(otherCardsStr.split(","));
                            pokerManager.threeMode = true;
                        }
                        break;
                    }
                    case names: {
                        String[] splitArray = recoverInfo.split("\\^");
                        int turnIndex = Integer.parseInt(splitArray[0]);
                        String namesStr = splitArray[1];

                        players = namesStr.split(",");
                        playInfo.name = players[turnIndex];
                        playInfo.turnIndex = turnIndex;
                        if(splitArray.length > 2) {
                            String threeNamesStr = splitArray[2];
                            threeModePlayers = threeNamesStr.split(",");

                            for(int i=0;i<4;++i) {
                                if(threeModePlayers[i].equals(playInfo.name)) {
                                    playInfo.turnIndex = i;
                                    AlertView.showViewSetText(context,String.format("You are player %d",i+1));
                                }

                                if ( threeModePlayers[i].equals("-")) {
                                    pokerManager.comIndex = i;
                                }
                            }
                        }
                        break;
                    }
                    case callRecord: {
                        String[] splitArray = recoverInfo.split("\\|");

                        for(int i=0;i<splitArray.length;++i) {
                            String player = splitArray[i];
                            if(player.equals("")) {
                                continue;
                            }
                            String[] records = player.split("\\^");
                            for(String card : records) {
                                if(!card.equals("")) {
                                    pokerManager.callsRecord.get(i).add(Integer.parseInt(card));
                                }
                            }
                        }
                        break;
                    }
                    case playRecord: {
                        String[] splitArray = recoverInfo.split("\\|");

                        for(int i=0;i<splitArray.length;++i) {
                            String player = splitArray[i];
                            if(player.equals("")) {
                                continue;
                            }
                            String[] records = player.split("\\^");
                            for(String card : records) {
                                if(!card.equals("")) {
                                    int poker = Integer.parseInt(card);
                                    int flower = (poker-1)/13;
                                    pokerManager.playsRecord.get(i).add(poker);
                                    pokerManager.flowerCountRecord[flower]+=1;
                                 }
                            }
                        }
                        break;
                    }
                    case boutsWin: {
                        String[] splitArray = recoverInfo.split("\\|");
                        int lastWinIndex = -1;
                        for (String bounts : splitArray) {
                            if (!bounts.equals("")) {
                                int winIndex = Integer.parseInt(bounts);
                                lastWinIndex = winIndex;
                                pokerManager.boutsWinRecord.add(winIndex);
                                if (winIndex == playInfo.turnIndex ||
                                        winIndex == (playInfo.turnIndex + 2) % 4) {
                                    pokerManager.ourScroe += 1;
                                } else {
                                    pokerManager.enemyScroe += 1;
                                }
                            }
                        }
                        if (lastWinIndex != -1 && pokerManager.boutsWinRecord.size() < pokerManager.playsRecord.get(lastWinIndex).size()) {
                            int playRecordWinSize = pokerManager.playsRecord.get(lastWinIndex).size();
                            pokerManager.currentFlower = ((pokerManager.playsRecord.get(lastWinIndex).get(playRecordWinSize - 1)));
                        }
                        break;
                    }
                    case done: {
                        String[] splitArray = recoverInfo.split("\\^");
                        GameState tempState = GameState.fromInteger(Integer.parseInt(splitArray[0]));
                        gameState = tempState;
                        pokerManager.turnIndex = Integer.parseInt(splitArray[1]);
                        pokerManager.trump = Integer.parseInt(splitArray[2]);
                        int callLast = Integer.parseInt(splitArray[3]);

                        if(pokerManager.turnIndex == -1 && playInfo.turnIndex == 0) {
                            new Thread(new Runnable() {
                                @Override
                                public void run() {
                                    try {
                                        sleep(1000);
                                        streamManager.sendMessage("P-1,0");
                                    } catch (InterruptedException e) {
//                                        Log.d("error");
                                    }
                                }
                            }).start();
                        }

                        if (!pokerManager.threeMode) {
                            if (callLast == playInfo.turnIndex || callLast == (playInfo.turnIndex + 2) % 4) {
                                pokerManager.winNumber = PokerManager.totalSum - (pokerManager.trump / 7 + 7);
                            } else {
                                pokerManager.winNumber = (pokerManager.trump / 7 + 7);
                            }
                        }
                        else {
                            if ( playInfo.turnIndex==0 || playInfo.turnIndex==2) {
                                pokerManager.winNumber = (pokerManager.trump / 7 + 7);
                            }
                            else {
                                pokerManager.winNumber = PokerManager.totalSum - (pokerManager.trump / 7 + 7);
                            }
                        }

                        if (listener != null) {
                            listener.changeStateUI(gameState);
                        }

                        if(gameState == GameState.Call) {
                            if(callListener != null) {
                                callListener.updateCallingUI(pokerManager.trump);
                            }
                        }
                        else if (gameState == GameState.CallChoosePartner) {
                            if(callListener != null) {
                                callListener.showThreeModeUI(pokerManager.turnIndex);
                            }
                        }
                        else if (gameState == GameState.Play) {
                            if(playListener != null) {
                                playListener.recoverPlayingUI();
                            }
                        }

                        break;
                    }
                }
                break;
            }
        }
    }

    @Override
    public void connectInterrupt() {
        if(StateManager.getInstance().isGameOver) {
            return;
        }

        if(gameState == GameState.Connected || gameState == GameState.Wait) {
            closeConnectAlert("Someone Leave");
        }
        else {
            AlertDialog.Builder alert = new AlertDialog.Builder(context);
            alert.setTitle("Server Message");
            alert.setMessage("Someone Disconnect");
            alert.setPositiveButton("OK", new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialogInterface, int i) {
                    exit(1); }});
            alert.show();
        }
    }

    public void closeConnectAlert(String message) {
        AlertDialog.Builder alert = new AlertDialog.Builder(context);
        alert.setTitle("Server Message");
        alert.setMessage(message);
        alert.setPositiveButton("OK", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                exit(1);
            }
        });
        alert.show();
    }

}

File: iOS/FancyBridge/Managers/StateManager.swift
//
//  StateManager.swift
//  FancyBridge
//
//  Created by Talka_Ying on 2018/1/24.
//  Copyright © 2018年 Talka_Ying. All rights reserved.
//

import UIKit

enum GameState:Int {
    case Connected = -1
    case Wait
    case Call
    case CallChoosePartner
    case Play
}

enum ConnectState:String {
    case GameState = "S"
    case Message = "M"
    case Wait = "W"
    case ThreeMode = "T"
    case Names = "N"
    case Hand = "H"
    case Call = "C"
    case Play = "P"
    case Disconnect = "D"
    case Recover = "R"
}

enum RecoverState:String {
    case cards = "1"
    case names = "2"
    case callRecord = "3"
    case playRecord = "4"
    case boutsWin = "5"
    case done = "6"
}

enum PlayState:Int {
    case Normal = 0
    case LastCard
    case GameOver
}

struct PlayerInfo {
    var name:String
    var turnIndex:Int
}

protocol StateManagerDelegate:class {
    func updateWaitingUI(_ normalCount:Int, _ threeModeCount:Int)
    func changeStateUI(_ state:GameState)
}

protocol StateManagerCallDelegate:class {
    func updateCallingUI(_ index:Int)
    func showThreeModeUI(_ index:Int)
}

protocol StateManagerPlayDelegate:class {
    func updatePlayingUI(_ poker:Int,_ type:Int,_ lastUser:Int)
    func recoverPlayingUI()
}

class StateManager: NSObject, StreamManagerDelegate {

    static var instance:StateManager!
    //MARK: - Member
    weak var delegate:StateManagerDelegate!
    weak var callDelegate:StateManagerCallDelegate!
    weak var playDelegate:StateManagerPlayDelegate!
    var streamManager:StreamManager!
    var playInfo:PlayerInfo = PlayerInfo(name: "", turnIndex: -1)
    var players:[String] = [String]()
    var threeModePlayers:[String] = [String]()
    var gameState:GameState = .Wait
    var isGameOver:Bool = false
    
    //MARK: - Init
    static func getInstance() -> StateManager {
        
        if instance == nil {
            instance = StateManager()
        }
        return instance
    }
    
    override init() {
        super.init()
        streamManager = StreamManager.getInstance()
        streamManager.delegate = self
    }
    
    //MARK: - Functions
    func reset() {
        players = [String]()
        gameState = .Wait
        isGameOver = false
        if (self.delegate != nil) {
            self.delegate.changeStateUI(gameState)
        }
    }
    
    func connectServer() {
        streamManager.connect()
    }
    
    func joinThreeMode() {
        streamManager.sendMessage("W")
    }
    
    func choosedPartner(_ choosedIndex:Int) {
        streamManager.sendMessage(String(format:"T%d",choosedIndex))
    }
    
    func interruptConnect() {
        streamManager.disconnect()
    }
    
    func test(_ text:String) {
        streamManager.sendMessage(text)
    }
    
    func getData(_ data: String) {
//        print(data)
        let pokerManager = PokerManager.getInstance()
        let connectState:ConnectState? = ConnectState(rawValue: String(data.first!))
        let info = String(data.dropFirst(1))
        
        if connectState == nil {
            return
        }
        
        switch connectState! {
        case .GameState:
            let splitArray = info.components(separatedBy: ",")
            let tempState = GameState(rawValue: Int(splitArray.first!)!)
            
            if tempState == nil {
                return
            }
            
            gameState = tempState!
            
            switch gameState {
            case .Connected:
                let DEVICE_UUID:String = UIDevice.current.identifierForVendor?.uuidString ?? UUID().uuidString
                streamManager.sendMessage("N"+playInfo.name+",I-"+DEVICE_UUID)
                break
            case .Wait:
                playInfo.turnIndex = Int(splitArray.last!)!
                break
            case .Call:
                let threeModeInt = Int(splitArray[2])!
                pokerManager.turnIndex = Int(splitArray[1])!
                pokerManager.threeMode = (threeModeInt==1 ? true : false)
                
                break
            case .CallChoosePartner:
                break
            case .Play:
                var lastUser = Int(splitArray.last!)!
                pokerManager.trump = Int(splitArray[1])!
                pokerManager.turnIndex = Int(splitArray[2])!
                pokerManager.callsRecord[lastUser].append(-1)
                
                if pokerManager.threeMode {
                    lastUser = 1
                }
                
                if (lastUser==playInfo.turnIndex||lastUser==(playInfo.turnIndex+2)%4) {
                    pokerManager.winNumber = PokerManager.totalSum-(pokerManager.trump/7+7)
                }
                else {
                    pokerManager.winNumber = (pokerManager.trump/7+7)
                }
                
                break
            }
            
            if (self.delegate != nil) {
                self.delegate.changeStateUI(gameState)
            }
            
            break
        case .Message:
//            if (self.delegate != nil) {
//                self.delegate.updateWaitingUI(info)
//            }
            break
        case .Wait:
            let splitArray = info.components(separatedBy: ",")
            let normalCount = Int(splitArray.first!)!
            let threeModeCount = Int(splitArray.last!)!
            
            if (self.delegate != nil) {
                self.delegate.updateWaitingUI(normalCount,threeModeCount)
            }
            break
        case .ThreeMode:
            let calledIndex = Int(info)!
            gameState = .CallChoosePartner
            pokerManager.turnIndex = calledIndex
            if (self.delegate != nil) {
                self.callDelegate.showThreeModeUI(calledIndex)
            }
            break
        case .Names:
            if !pokerManager.threeMode {
                players = info.components(separatedBy: ",")
            }
            else {
                threeModePlayers = info.components(separatedBy: ",")

                for i in 0..<4 {
                    if threeModePlayers[i] == playInfo.name {
                        playInfo.turnIndex = i
                        AlertView.showViewSetText(String(format:"You are player %d",i+1))
                        break
                    }
                }
                pokerManager.comIndex = threeModePlayers.index(of: "-")!
            }
            break
        case .Hand:
            if !pokerManager.threeMode {
                pokerManager.setCards(info.components(separatedBy: ","))
            }
            else {
                pokerManager.setOtherCards(info.components(separatedBy: ","))
            }
            break
        case .Call:
            let splitArray = info.components(separatedBy: ",")
            let nowTurn = Int(splitArray[1])!
            let trump = Int(splitArray.last!)!
            let lastTurn = Int(splitArray.first!)!
            
            pokerManager.turnIndex = nowTurn
            if playInfo.turnIndex == nowTurn {
                AlertView.showViewSetText("Your Turn")
            }
            
            pokerManager.callsRecord[lastTurn].append(trump)
            
            if (splitArray.last != nil) {
                if (self.callDelegate != nil) {
                    self.callDelegate.updateCallingUI(trump)
                }
            }
            break
        case .Play:
            let splitArray = info.components(separatedBy: ",")
            let nextTurn = Int(splitArray.last!)!
            let lastTurn = (nextTurn+4-1)%4;
            let poker = Int(splitArray.first!)!
            let playState = PlayState(rawValue: Int(splitArray[1])!)!
            
            pokerManager.turnIndex = nextTurn
            
            switch playState {
            case .Normal:
                if poker != 0 {
                    if pokerManager.currentFlower == -1 {
                        pokerManager.currentFlower=(poker-1)/13;
                    }
                    
                    let flower = Int((poker-1)/13)
                    pokerManager.playsRecord[lastTurn].append(poker)
                    pokerManager.flowerCountRecord[flower] += 1;
                }
                else {
                    pokerManager.boutsWinRecord.append(nextTurn)
                }
                break
            case .LastCard:
               
                let flower = Int((poker-1)/13)
                pokerManager.playsRecord[lastTurn].append(poker)
                pokerManager.flowerCountRecord[flower] += 1;
                
                if playInfo.turnIndex == 0 {
                    DispatchQueue.global().async {
                        sleep(2)
                        self.streamManager.sendMessage("P-1,0")
                    }
                }
                
                pokerManager.currentFlower = -1
                pokerManager.turnIndex = -1
                
                break
            case .GameOver:
                pokerManager.boutsWinRecord.append(nextTurn)
                isGameOver = true
                break
            }
            
            if (self.playDelegate != nil) {
                self.playDelegate.updatePlayingUI(poker, playState.rawValue, lastTurn)
            }
            
            break
        case .Disconnect:
            if StateManager.getInstance().isGameOver {
                return
            }
            
            if info == "0" {
                UIAlertController.showAlert(title: "Server Message", message: "Someone Disconnect")
                LoadingView.StartLoading()
            }
            else if info == "1" {
                if (gameState == .CallChoosePartner && pokerManager.turnIndex == playInfo.turnIndex)
                    || gameState != .CallChoosePartner {
                    LoadingView.StopLoading()
                }
            }
            else {
                closeConnectAlert("close connect")
            }
            
            break
        case .Recover:
            let recoverState:RecoverState? = RecoverState(rawValue: String(info.first!))
            let recoverInfo = String(info.dropFirst(2))
            
            switch recoverState {
            case .cards:
                let splitArray = recoverInfo.components(separatedBy: "^")
                let cards = splitArray[0]
                pokerManager.setCards(cards.components(separatedBy: ","))
                if splitArray.count > 1 {
                    let otherCards = splitArray[1]
                    pokerManager.setOtherCards(otherCards.components(separatedBy: ","))
                    pokerManager.threeMode = true
                }
                break
            case .names:
                let splitArray = recoverInfo.components(separatedBy: "^")
                let turnIndex = Int(splitArray[0])!
                let names = splitArray[1]
                
                players = names.components(separatedBy: ",")
                playInfo.name = players[turnIndex]
                playInfo.turnIndex = turnIndex
                if splitArray.count > 2 {
                    let threeNames = splitArray[2]
                    threeModePlayers = threeNames.components(separatedBy: ",")
                    
                    for i in 0..<4 {
                        if threeModePlayers[i] == playInfo.name {
                            playInfo.turnIndex = i
                            AlertView.showViewSetText(String(format:"You are player %d",i+1))
                            break
                        }
                    }
                    pokerManager.comIndex = threeModePlayers.index(of: "-")!
                }
                break
            case .callRecord:
                let splitArray = recoverInfo.components(separatedBy: "|")

                for i in 0..<4 {
                    let player = splitArray[i]
                    if player == "" {
                        continue
                    }
                    let records = player.components(separatedBy: "^")
                    for card in records {
                        if card != "" {
                            pokerManager.callsRecord[i].append(Int(card)!)
                        }
                    }
                }
                break
            case .playRecord:
                let splitArray = recoverInfo.components(separatedBy: "|")

                for i in 0..<4 {
                    let player = splitArray[i]
                    if player == "" {
                        continue
                    }
                    let records = player.components(separatedBy: "^")
                    for card in records {
                        if card != "" {
                            let poker = Int(card)!
                            let flower = Int((poker-1)/13)
                            pokerManager.playsRecord[i].append(poker)
                            pokerManager.flowerCountRecord[flower] += 1;
                        }
                    }
                }
                break
            case .boutsWin:
                let splitArray = recoverInfo.components(separatedBy: "|")
                var lastWinIndex = -1
                for bouts in splitArray {
                    if bouts != "" {
                        let winIndex = Int(bouts)!
                        lastWinIndex = winIndex
                        pokerManager.boutsWinRecord.append(winIndex)
                        if (winIndex==playInfo.turnIndex||winIndex==(playInfo.turnIndex+2)%4) {
                            pokerManager.ourScroe += 1
                        }
                        else {
                            pokerManager.enemyScroe += 1
                        }
                    }
                }
                if lastWinIndex != -1 && pokerManager.boutsWinRecord.count < pokerManager.playsRecord[lastWinIndex].count {
                    pokerManager.currentFlower=((pokerManager.playsRecord[lastWinIndex].last!)-1)/13;
                }
                
                break
            case .done:
                let splitArray = recoverInfo.components(separatedBy: "^")
                let tempState = GameState(rawValue: Int(splitArray[0])!)!
                gameState = tempState
                pokerManager.turnIndex = Int(splitArray[1])!
                pokerManager.trump = Int(splitArray[2])!
                let callLast = Int(splitArray[3])!
                
                if pokerManager.turnIndex == -1 && playInfo.turnIndex == 0 {
                    DispatchQueue.global().async {
                        sleep(1)
                        self.streamManager.sendMessage("P-1,0")
                    }
                }
                
                if (!pokerManager.threeMode) {
                    if (callLast==playInfo.turnIndex||callLast==(playInfo.turnIndex+2)%4) {
                        pokerManager.winNumber = PokerManager.totalSum-(pokerManager.trump/7+7)
                    }
                    else {
                        pokerManager.winNumber = (pokerManager.trump/7+7)
                    }
                }
                else {
                    if ( playInfo.turnIndex==0 || playInfo.turnIndex==2) {
                        pokerManager.winNumber = (pokerManager.trump / 7 + 7);
                    }
                    else {
                        pokerManager.winNumber = PokerManager.totalSum - (pokerManager.trump / 7 + 7);
                    }
                }
                
                if (self.delegate != nil) {
                    self.delegate.changeStateUI(gameState)
                }
                
                if gameState == .Call {
                    if (self.callDelegate != nil) {
                        self.callDelegate.updateCallingUI(pokerManager.trump)
                    }
                }
                else if gameState == .CallChoosePartner {
                    if (self.callDelegate != nil) {
                        self.callDelegate.showThreeModeUI(pokerManager.turnIndex)
                    }
                }
                else if gameState == .Play {
                    if (self.playDelegate != nil) {
                        self.playDelegate.recoverPlayingUI()
                    }
                }

                break
            case .none:
                break
            }
            
            break
        }
    }
    
    func connectInterrupt() {
        if StateManager.getInstance().isGameOver {
            return
        }
        
        if gameState == .Connected || gameState == .Wait {
            closeConnectAlert("Someone Leave")
        }
        else {
            let cancelAction:UIAlertAction = UIAlertAction(title: "OK", style: .default, handler: { (UIAlertAction) in
                exit(1)
                })
            UIAlertController.showAlert(title: "Server Message",message: "Someone Disconnect",actions: [cancelAction])
        }
    }
    
    func closeConnectAlert(_ message:String) {
        let cancelAction:UIAlertAction = UIAlertAction(title: "OK", style: .default, handler: { (UIAlertAction) in
                exit(1)
            })
        
        UIAlertController.showAlert(title: "Server Message",message: message,actions: [cancelAction])
    }
}
